### ISSUE TITLE
Non-Static Lambdas in `on_submit` et al.

### ISSUE THREAD

**Issue by hemangandhi** [gyscos/Cursive#179]
The following is an attempt to get a formatted string from the user. `out` is the validator function and `and_then` is the action to take on valid input.

```rust
    pub fn prompt_user<Ret, F, S, A>(prompt: S, siv: &mut Cursive, out: &F, and_then: &A)
    where F: Fn(&str) -> Result<Ret, &str> + 'static, S: Into<String> + 'static,
          A: Fn(&mut Cursive, Ret) + 'static{
        siv.add_layer(Dialog::around(EditView::new()
                                        .on_submit(move |s, txt| {
            match out(txt) {
                Ok(r) => and_then(s, r),
                Err(es) => prompt_user(prompt, s, out, and_then)
            }
        })))
    }
```

Currently, this doesn't compile: the lambda in `on_submit` cannot be `'static`. Hence, I was wondering if there was a better way to get this sort of functionality. Or, perhaps, could `EditView` accept non-static lambdas (I presume the lambda would have to outlive the view, but that's all, right)?

More generally, is there a reason the lambda has to be static?

**Comment by gyscos** [comment]
Hi!

Accepting non-`'static` closures would be pretty difficult: as you guessed, the lambda must (at least) outlive the view, but the view itself is also `static` (there is no lifetime tied to the view).

In your case, there is no guarantee that either `and_then` or `out` will still exist after `prompt_user` returns; so we cannot use them from the callback, which could easily last longer that that (maybe as long as the `Cursive` itself exists).

In addition, the callback may be fired multiple times (we can't prove otherwise at compilation time), so it cannot consume its environment: you cannot send `prompt` as-is, because `prompt_user` consumes it.
The idea is the same for `and_then` and `out`: they have to exist somewhere (to be "owned" by something), but here the ownership model is not as clear.

This is a common problem with a few classical solutions:
* If you're certain the callback will only be called once, you can "give up" ownership of the variables to the callback. The callback itself is a `FnMut`, not `FnOnce`, so to work around that, you can use a trick involving some `Option`s: the callback will modify its environment; the first time, it will empty the options and use them as expected; but if called a second time, it will panic:

```
pub fn prompt_user<Ret, F, S, A>(
    prompt: S, siv: &mut Cursive, out: Rc<F>, and_then: A
) where
    F: Fn(&str) -> Result<Ret, &str> + 'static,
    S: Into<String> + 'static,
    A: Fn(&mut Cursive, Ret) + 'static,
{
    let mut and_then = Some(and_then);
    let mut out = Some(out);
    let mut prompt = Some(prompt);

    siv.add_layer(Dialog::around(EditView::new().on_submit_mut(
        move |s, txt| {
            let prompt = prompt.take().unwrap();
            let out = out.take().unwrap();
            let and_then = and_then.take().unwrap();

            match out(txt) {
                Ok(r) => and_then(s, r),
                Err(es) => prompt_user(prompt, s, out, and_then),
            }
        },
    )))
}
```

* Clone everything. If everything is going to be small and light (closures with almost no environment, small strings), cloning it is a good way to give each component what it needs.
```
pub fn prompt_user<Ret, F, S, A>(
    prompt: S, siv: &mut Cursive, out: F, and_then: A
) where
    F: Clone + Fn(&str) -> Result<Ret, &str> + 'static,
    S: Clone + Into<String> + 'static,
    A: Clone + Fn(&mut Cursive, Ret) + 'static,
{
    siv.add_layer(Dialog::around(EditView::new().on_submit(move |s, txt| {
        match out(txt) {
            Ok(r) => and_then(s, r),
            Err(es) => {
                prompt_user(prompt.clone(), s, out.clone(), and_then.clone())
            }
        }
    })))
}
```
* Shared ownership: if you don't want or can't clone the arguments, you will need to share ownership, with a `Rc` for instance (this is closer to what languages with GC do).

```
pub fn prompt_user<Ret, F, S, A>(
    prompt: Rc<S>, siv: &mut Cursive, out: Rc<F>, and_then: Rc<A>
) where
    F: Fn(&str) -> Result<Ret, &str> + 'static,
    S: Into<String> + 'static,
    A: Fn(&mut Cursive, Ret) + 'static,
{
    siv.add_layer(Dialog::around(EditView::new().on_submit(move |s, txt| {
        match out(txt) {
            Ok(r) => and_then(s, r),
            Err(es) => {
                prompt_user(prompt.clone(), s, out.clone(), and_then.clone())
            }
        }
    })))
}
```

The general idea is:
* You need to make sure everything the closure uses will be alive when it is called. References are too risky, as there is no guarantee they'll still be valid.
* Callbacks can (theorically) be called multiple times. This means they cannot consume their environment. What they *can* do is mutate it: if you want to consume a `T` and know it'll only get called once, you can keep a `Option<T>` in the environment, and use `Option::take` to retrieve its value:


This last solution is really expecting that the callback will not be called twice, otherwise it will panic.

**Comment by sandeep-datta** [comment]
> but the view itself is also static (there is no lifetime tied to the view).

Pardon me, my rust is (ahem) a bit rusty, but does the view have to be static? Say for example can't we thread the life-time parameter from `Dialog` to `ChildButton` to `Button` to button callback? IMO the static lifetime is a little too restrictive and the workarounds all have some draw back.

**Comment by gyscos** [comment]
Lifetimes are checked at compile-time. In this case, this means that views have (at least) the same lifetime as the Cursive root itself, since we can't prove _when_ the view will be removed.

I considered at some point removing this static lifetime bound, but it would mean adding a lifetime parameter to the Cursive root, and almost every other struct.
I'm not convinced yet this is worth what it would bring - basically just being able to reference variables defined before `Cursive` itself. In these cases a Rc could be a much easier solution.

**Comment by sandeep-datta** [comment]
> I'm not convinced yet this is worth what it would bring - basically just being able to reference variables defined before Cursive itself. In these cases a Rc could be a much easier solution.

I agree Rc is an easier solution but as I am sure you already know it's less efficient than being able to use plain references. And it is also less than perfect since it can leak memory (if there are cyclic dependencies). 

I am a new user but I have a feeling this library is going to play a foundational role in Rust's command line story (if it's not doing that already). So IMHO Cursive being a library should try and avoid making this decision for its end users. Let's give the end users the best Rust has to offer.

**Comment by gyscos** [comment]
`Rc` ofter perform better than people expect - I wouldn't consider it a performance issue without actual benchmarks. It has no lock, no atomic transaction, and in regular use is just as efficient as a reference (or a `Box`). Cloning and dropping do add a checked addition or subtraction, but that's far from a significant impact. If you just use it where you would use a reference, it cannot lead to cycles and won't cause much trouble.

`Arc` adds a bit more overhead for Clones/Drops (from the atomic operations), but that's still only rarely an issue, unless you're cloning it thousands of times per second - in which case you may have other problems.

Do you have a use-case where this is currently a problem? I may be able to help get an ergonomic solution :)
